scripts/build-release.sh

```diff
@@ -0,0 +1,131 @@
+
+# Arguments
+RELEASE_TAG="${1-}"
+
+# Globals
+QUAY_REPO="${QUAY_REPO:-kubernetes-multicluster/kubefed}"
+GITHUB_REPO="${GITHUB_REPO:-kubernetes-sigs/kubefed}"
+GITHUB_REMOTE_UPSTREAM_NAME="${GITHUB_REMOTE_UPSTREAM_NAME:-upstream}"
+
+function verify-command-installed() {
+  if ! util::command-installed travis; then
+    echo "travis command not found. Please add travis to your PATH and try again." >&2
+    return 1
+  fi
+}
+
+function build-release-artifacts() {
+  ${ROOT_DIR}/scripts/build-release-artifacts.sh ${RELEASE_TAG}
+}
+
+function create-and-push-tag() {
+  # Use the upstream git remote convention name used by hub.
+  # This creates an annotated tag required to ensure that the KubeFed binaries
+  # are versioned correctly.
+  git tag -a "${RELEASE_TAG}" "${GITHUB_REMOTE_UPSTREAM_NAME}/master" -m "Creating release tag ${RELEASE_TAG}"
+  git push "${GITHUB_REMOTE_UPSTREAM_NAME}" "${RELEASE_TAG}"
+}
+
+function travis-build-status() {
+  local status="$(travis branches --repo ${GITHUB_REPO} | grep "${RELEASE_TAG}" | awk '{print $3}')"
+  echo "${status}"
+}
+
+function verify-travis-build-started() {
+  local buildStatus="$(travis-build-status)"
+  [[ "${buildStatus}" == "started" ]]
+}
+
+function verify-travis-build-completed() {
+  local statuses="${passedStatus} failed errored canceled"
+  local buildStatus="$(travis-build-status)"
+
+  for status in ${statuses}; do
+    if [[ "${status}" == "${buildStatus}" ]]; then
+      return 0
+    fi
+  done
+
+  return 1
+}
+
+function verify-continuous-integration() {
+  local passedStatus="passed"
+  util::wait-for-condition "kubefed CI build to start" "verify-travis-build-started started" 300
+  util::wait-for-condition "kubefed CI build to complete" "verify-travis-build-completed" 3600
+  local buildStatus="$(travis-build-status)"
+
+  if [[ "${buildStatus}" == "${passedStatus}" ]]; then
+    echo "kubefed CI build ${passedStatus}"
+  else
+    echo "kubefed CI build ${buildStatus}. Exiting."
+    exit 1
+  fi
+}
+
+function quay-image-status() {
+  local quayImagesApi="https://quay.io/api/v1/repository/${QUAY_REPO}/image/"
+  curl -s ${quayImagesApi} | grep "${RELEASE_TAG}" &> /dev/null
+}
+
+function verify-container-image() {
+  util::wait-for-condition "kubefed container image in quay" "quay-image-status" 60
+}
+
+function update-changelog() {
+  sed -i "/# Unreleased/a \\\n# ${RELEASE_TAG}" CHANGELOG.md
+}
+
+RELEASE_TAG_REGEX="^v[0-9]+\.[0-9]+\.[0-9]+(-rc[0-9]+)?$"
+if [[ ! "${RELEASE_TAG}" =~ ${RELEASE_TAG_REGEX} ]]; then
+  >&2 echo "usage: $0 <release tag of the form v[0-9]+.[0-9]+.[0-9]+(-rc[0-9]+)?>"
+  exit 1
+fi
+
+util::log "Verifying travis CLI command installed"
+verify-command-installed
+
+util::log "Building release artifacts first to make sure build succeeds"
+build-release-artifacts
+
+util::log "Creating local git annotated tag and pushing tag to kick off build process"
+create-and-push-tag
+
+util::log "Verifing image builds and completes successfully in Travis"
```

Comment:
Maybe indicate in the log message that this is going to take a really long time (e.g. ~60s)?  Also, what happens if the user accidentally cancels the script?  Could it optionally restart at the verifying step?

Reply:
Er, I meant 60m.